### PR TITLE
[LC-193] Add KeyValueStore instead of leveldb

### PR DESCRIPTION
**Pull request by hangaroy**
- Change database library from leveldb to plyvel.
- No update python [leveldb](https://pypi.org/project/leveldb/0.20/) client since 0.20.
- remove kev_value_store_factory
- refactoring key-value store close logic
- fixed to specifies keyword-only parameter
- update unittest to test all store(dict, leveldb, plyvel)

### CODE REVIEW

**Review by windies21 — APPROVED** [review]

**Comment by zsaladin** [comment]
Add description please.

File: loopchain/store/key_value_store.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    def CancelableWriteBatch(self, sync=False) -> KeyValueStoreCancelableWriteBatch:
+        raise NotImplementedError("CancelableWriteBatch constructor is not implemented in KeyValueStore class")
+
+    """Return iterator
+
+    :param start_key: a start key (inclusive)
+    :param stop_key: a stop key (inclusive)
+    :param include_value: include value (for key, value in store_instance.Iterator(include_value=True):)
+    :return: iterator
+    """
```
Comment (metalg0su) [inline loopchain/store/key_value_store.py]: I think this docstring should be re-located or deleted

File: loopchain/store/key_value_store_factory.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        if store_type == KeyValueStoreFactory.STORE_TYPE_PLYVEL:
+            utils.logger.debug(f"New KeyValueStorePlyvel.")
+            return KeyValueStorePlyvel(uri, **kwargs)
+        elif store_type == KeyValueStoreFactory.STORE_TYPE_LEVELDB:
+            utils.logger.warning(f"New KeyValueStoreLevelDb. store_type={store_type}, uri={uri}")
+            return KeyValueStoreLevelDb(uri, **kwargs)
+        elif store_type == KeyValueStoreFactory.STORE_TYPE_DICT:
+            raise ValueError(f"Unsupport KeyValueStoreDict. If you want to use it, you need to remove this line")
+            return KeyValueStoreDict(uri, **kwargs)
```
Comment (metalg0su) [inline loopchain/store/key_value_store_factory.py]: This line is unreachable, if ValueError raised?

**Comment by hangaroy** [comment]
> Add description please.

Added

**Comment by hangaroy** [comment]
> Add description please.

File: requirements.txt [hunk 3]
```diff
@@ -22,3 +22,4 @@ verboselogs==1.7
 websockets==6.0
 yappi==0.98
 dataclasses; python_version < '3.7'
+plyvel==1.0.5
```
Comment (yakkle) [inline requirements.txt]: remove leveldb in requirements.txt

**Review by metalg0su — APPROVED** [review]

File: loopchain/store/key_value_store.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+import abc
+import functools
+
+
+class KeyValueStoreError(Exception):
+    pass
+
+
+class KeyValueStoreWriteBatch(metaclass=abc.ABCMeta):
```
Comment (yakkle) [inline loopchain/store/key_value_store.py]:
```suggestion
class KeyValueStoreWriteBatch(abc.ABC):
```

File: loopchain/store/key_value_store.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if exc_type is not None:
+            self.write()
+
+
+class KeyValueStoreCancelableWriteBatch(metaclass=abc.ABCMeta):
```
Comment (yakkle) [inline loopchain/store/key_value_store.py]:
```suggestion
class KeyValueStoreCancelableWriteBatch(abc.ABC):
```

File: loopchain/store/key_value_store.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    def _touch(self, key: bytes):
+        raise NotImplementedError("_touch() function is interface method")
+
+    @abc.abstractmethod
+    def _get_original_touched_item(self):
+        # Children have to override function as generator. return key, value
+        raise NotImplementedError("_get_touched_item() function is interface method")
+
+
+class KeyValueStore(metaclass=abc.ABCMeta):
```
Comment (yakkle) [inline loopchain/store/key_value_store.py]:
```suggestion
class KeyValueStore(abc.ABC):
```

File: requirements.txt (hunk 3 above)
Comment (yakkle) [reply requirements.txt]: leveldb supported in key_value_store_leveldb. I resolve this comment.

Comment (hangaroy) [reply requirements.txt]: > remove leveldb in requirements.txt

Please see key_value_store_leveldb.py, test_key_value_store.py, key_value_store_factory.py
KeyValueStore still support leveldb library.
We can use leveldb library as KeyValueStore backend library if we want.

**Review by yakkle — APPROVED** [review]

File: loopchain/peer/block_manager.py [hunk 7]
```diff
@@ -53,20 +54,19 @@ class BlockManager:
...
-        self.__level_db = None
-        self.__level_db_path = ""
-        self.__level_db, self.__level_db_path = util.init_level_db(
-            level_db_identity=f"{level_db_identity}_{channel_name}",
-            allow_rename_path=False
-        )
+
+        key_value_store, key_value_store_path = util.init_default_key_value_store(f"{store_identity}_{channel_name}")
+        self.__key_value_store = key_value_store
+        self.__key_value_store_path = key_value_store_path
```
Comment (zsaladin) [inline loopchain/peer/block_manager.py]: Managing `key_value store` in two classes is not nessesary.
`get_key_value_store` has to return `key_value_store` from `blockchain` class.

File: loopchain/peer/block_manager.py [hunk 8]
```diff
@@ -737,14 +737,14 @@ def __get_peer_stub_list(self):
...
+    def __close_key_value_store(self):
+        self.__key_value_store.close()
+        self.__key_value_store: KeyValueStore = None
+        self.__blockchain.close_blockchain_store()
 
     def stop(self):
-        # for reuse level db when restart channel.
-        self.__close_level_db()
+        # for reuse key value store when restart channel.
+        self.__close_key_value_store()
```
Comment (zsaladin) [inline loopchain/peer/block_manager.py]: Then  `BlockManager` don't need to close KeyValueStore.
Just call it `blockchain.close()`

File: loopchain/store/key_value_store_dict.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+from loopchain.store.key_value_store import KeyValueStoreError
+from loopchain.store.key_value_store import KeyValueStoreWriteBatch, KeyValueStoreCancelableWriteBatch, KeyValueStore
+from loopchain.store.key_value_store import _validate_args_bytes, _validate_args_bytes_without_first
+
+
+def _error_convert(func):
+    @functools.wraps(func)
+    def _wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
```
Comment (zsaladin) [inline loopchain/store/key_value_store_dict.py]: Checking that value type is `bytes` is needed.

File: loopchain/store/key_value_store_dict.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+            _validate_args_bytes(default)
+
+        try:
+            return self._store_items[key]
+        except KeyError:
+            if default is None:
+                raise KeyError(f"Has no value of key({key}")
+            return default
+
+    @_validate_args_bytes_without_first
```
Comment (zsaladin) [inline loopchain/store/key_value_store_dict.py]: Passing parameter explicitly sync=True raises an exception.

File: loopchain/blockchain/blockchain.py [hunk 11]
```diff
@@ -74,14 +75,17 @@ def __init__(self, blockchain_db=None, channel_name=None):
         self.__peer_id = ChannelProperty().peer_id
 
         # block db has [ block_hash - block | block_height - block_hash | BlockChain.LAST_BLOCK_KEY - block_hash ]
-        self.__confirmed_block_db = blockchain_db
-        # logging.debug(f"BlockChain::init confirmed_block_db({self.__confirmed_block_db})")
+        self.__confirmed_block_store = blockchain_store
```
[the remark this answers is not in the document]
Comment (yakkle) [reply loopchain/blockchain/blockchain.py]: 👌

File: loopchain/store/key_value_store_factory.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+from loopchain.store.key_value_store_plyvel import KeyValueStorePlyvel
+
+
+class KeyValueStoreFactory:
+    STORE_TYPE_PLYVEL = 'plyvel'
+    STORE_TYPE_LEVELDB = 'leveldb'
+    STORE_TYPE_DICT = 'dict'
+
+    @staticmethod
+    def new(uri: str, store_type: str=None, **kwargs) -> KeyValueStore:
```
[the remark this answers is not in the document]
Comment (yakkle) [reply loopchain/store/key_value_store_factory.py]: 👌

File: loopchain/store/key_value_store_dict.py (hunk 10 above)
Comment (yakkle) [reply loopchain/store/key_value_store_dict.py]: I don't know how did you see exception raised.
Can you explain more detail?

**Review by zsaladin — APPROVED** [review]

**Pull request merged**